### DISCUSSION TITLE
what does "Non-Java Unknown (native) Unknown (Java) Not walkable (Java)" mean in profile result

### DISCUSSION THREAD

**Discussion by cyflhn** [async-profiler/async-profiler#409] [Q&A] [answered]
I can see "Non-Java Unknown (native) Unknown (Java) Not walkable (Java)" on the top of the profile result, but I don't understand what does these mean? Is there anyone who could explain to me about these items?

**Comment by apangin** [comment, marked as answer]
All these states: `unknown_not_Java`, `unknown_Java`, `not_walkable_Java` mean that the profiler was not able to get Java stack trace at the specific point.

Async-profiler samples a thread state asynchronously at an arbitrary moment of time (hence the work `async` in its name). In particular, this means that the profiler may attempt to get a sample at some transitional state when the execution stack is inconsistent (e.g. in the middle of constructing a new stack frame or in the middle of calling a native method from Java).

The specific message depends on the nuances of the stack walking algorithm:

- `unknown_not_Java` means the thread is running native (VM or JNI) code;
- `unknown_Java` means the thread is running Java code, but the JVM is unable to detect the top Java frame on the stack;
- `not_walkable_Java` means the JVM has found the top frame, but cannot walk to the next frame.

The difference between them probably makes more sense for the profiler developers rather than for end users. You may treat all of them as some transitional thread state when the exact stack trace is unknown.

Normally, the amount of such unknown frames should be small (<5%), and they can be simply ignored. If you happen to see a case with the large amount of unknown frames, please let me know.

**Reply by cyflhn** [reply]
We observe that sometimes the profile result shows the amount of unknown frames are larger than 5%. Here is an example:
Non-Java:              509 (11.41%)
Unknown (native):      773 (17.33%)
Unknown (Java):        4 (0.09%)
Not walkable (Java):   7 (0.16%)

**Reply by apangin** [reply]
Ah, I see now.
You are using too old version of async-profiler.
Here `Non-Java` and `Unknown (native)` include GC threads and JIT Compiler threads respectively. It's normal for GC and Compiler to consume that much CPU. Actual "unknowns" here are just `Unknown (Java)` and `Not walkable (Java)` with a tiny percentage, which is absolutely fine.

In newer versions, GC and JIT are excluded from "unknown" summary to avoid confusion, but they are still displayed in a regular profile as usual.

**Reply by reliefeai** [reply]
I got CPU profile flamegraph like this :
<img width="1206" alt="截屏2021-06-10 下午3 06 06" src="https://user-images.githubusercontent.com/9815635/121480193-6565a080-c9fd-11eb-8512-7ec176d96ce1.png">

I suspect it's cause by  Linux version, in this pod Linux version is: 
`Linux version 3.10.0-957.21.3.el7.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC) ) #1 SMP Tue Jun 18 16:35:19 UTC 2019`

In another pod(same docker image),  I got right CPU profile flamegraph, it's Linux version is:
`Linux version 3.10.0-1160.15.2.el7.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 4.8.5 20150623(Red Hat 4.8.5-44) (GCC) ) #1 SMP Wed Feb 3 15:06:38 UTC 2021`

Is it Linux version too low or something? Or some other preconditions before profiling CPU ?

**Reply by apangin** [reply]
@liuzhiguo630 I would suspect a "wrong" JVM version.

**Reply by reliefeai** [reply]
> @liuzhiguo630 I would suspect a "wrong" JVM version.

They are same docker image, and same JVM, I've changed from openJDK to Liberica JDK for profiling memory alloc: 
`LibericaJDK 64-Bit Server VM (25.212-b12) for linux-amd64 JRE (1.8.0_212-BellSoft-b12), built on Apr 18 2019 18:06:04 by "bellsoft" with gcc 7.3.1 20180303 (Red Hat 7.3.1-5)`

update:
I notice this warning:
```
WARNING: Kernel symbols are unavailable due to restrictions. Try
  echo 0 > /proc/sys/kernel/kptr_restrict
  echo 1 > /proc/sys/kernel/perf_event_paranoid
```

I try to changed, but it still not work, I can see more cpu method from system(or jvm? i'm not sure), but java methods are flat too(the green part):
<img width="1211" alt="截屏2021-06-10 下午4 41 20" src="https://user-images.githubusercontent.com/9815635/121493734-b0d27b80-ca0a-11eb-94f8-4fc91f0ec42a.png">

**Reply by apangin** [reply]
Well, the problem looks like the JVM's AsyncGetCallTrace is broken. If this is not a JVM bug, then this might be a kernel bug. I haven't seen this before, though. But I can guess this may happen if the OS does not fill the signal context properly.

What if you profile in `-e itimer` mode?

**Reply by reliefeai** [reply]
sry it took too long, event itimer not work either.
<img width="1241" alt="截屏2021-06-15 15 24 15" src="https://user-images.githubusercontent.com/9815635/122010487-c02c3d00-cded-11eb-9815-88613b5d7a6d.png">

**Reply by apangin** [reply]
@liuzhiguo630 Please open a separate issue, as this seems to be out of scope of the original question.
If the same test works fine on another system, please include relevant differences in the OS settings and the hardware,
like `sysctl -a`, `ulimit -a`, `cat /proc/cpuinfo`

**Reply by danielshaya** [reply]
Hi

Thx for this fantastic product!

I'm experiencing the same issue as above about half my traces end up in the unknown_java bucket. I wonder if it's because I'm running on Mac osx (M1).  Do have any experience with this?

Thanks

**Reply by apangin** [reply]
@danielshaya Please open a new issue and provide more details about OS and JVM version, profiler setup, an example program if possible, etc.

**Reply by danielshaya** [reply]
If you run the Aeron Archiving Media Driver for example you get this profile
<img width="2971" alt="Screenshot 2022-10-27 at 10 47 08" src="https://user-images.githubusercontent.com/5359230/198252555-cb16573a-1360-4cc9-bb38-00c11c375243.png">

When I run my own program about half the stack is unknown_java
I'm running with the Oracle 17 jdk :
java version "17.0.4" 2022-07-19 LTS
Java(TM) SE Runtime Environment (build 17.0.4+11-LTS-179)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.4+11-LTS-179, mixed mode, sharing)
(Also tried with openjdk -same result)

ProductName:	macOS
ProductVersion:	12.6
BuildVersion:	21G115

**Comment by ditogam** [comment]
Hello Andrei,
Thank you for amazing product, it helps a lot in development and performance testing.

I have posted question on stackoverflow, Can you give any advice .

[Java performance issue On Oracle Linux](https://stackoverflow.com/questions/68205982/java-performance-issue-on-oracle-linux)

Thanks in advance

**Comment by realxujiang** [comment]
Hi, I encountered some performance issues while running data processing on an ARM server for big data. I'm curious about the significance of the large number of 'no_java_frame' entries in the flame graph. Are there any further steps I can take to investigate this?
![image](https://github.com/async-profiler/async-profiler/assets/12759642/cddce9f5-fa89-4748-a3fc-2367190e9d32)

**Reply by apangin** [reply]
Please open a new issue / start new discussion.
Specify which versions of the JVM, OS and async-profiler you are using. What is the exact command line to start profiler?

**Reply by realxujiang** [reply]
Hi, I have opened a new issue. Could you please help take a look?   #893
